internal/server/data/data.go

```diff
@@ -219,6 +219,15 @@ func (t *Transaction) Rollback() error {
 	return err
 }
 
+// SoftCommit commits the transaction if it has not already been committed.
+// if it has been committed, it does not error.
+func (t *Transaction) SoftCommit() error {
+	if t.completed.Load() {
+		return nil
+	}
+	return t.Commit()
+}
```

Comment:
I suspect it reverts to auto-commit mode, but anywhere I commit I start a new transaction for the remainder of the request. The problem seems to be tying that new transaction into the original context/db as if it was the original transaction. That sounds problematic. This means if the handler wants to commit the transaction, there's not much choice here. Alternatively you could choose to avoid trying to commit or rollback at all, but I don't really like that: What if it fails before it gets to the commit?